src/cascadia/QueryExtension/GithubCopilotLLMProvider.cpp

```diff
@@ -0,0 +1,365 @@
+                errorType = ErrorTypes::Unknown;
+                break;
+            }
+
+            _refreshAuthTokens();
+            refreshAttempted = true;
+        }
+
+        // Also make a new entry in our jsonMessages list, so the AI knows the full conversation so far
+        WDJ::JsonObject responseMessageObject;
+        responseMessageObject.Insert(roleKey, WDJ::JsonValue::CreateStringValue(assistantKey));
+        responseMessageObject.Insert(contentKey, WDJ::JsonValue::CreateStringValue(message));
+        _jsonMessages.Append(responseMessageObject);
+
+        co_return winrt::make<GithubCopilotResponse>(message, errorType, RS_(L"GithubCopilot_ResponseMetaData"));
+    }
+
+    void GithubCopilotLLMProvider::_refreshAuthTokens()
+    {
+        WDJ::JsonObject jsonContent;
+        jsonContent.SetNamedValue(clientIdKey, WDJ::JsonValue::CreateStringValue(windowsTerminalClientID));
+        jsonContent.SetNamedValue(grantTypeKey, WDJ::JsonValue::CreateStringValue(refreshTokenKey));
+        jsonContent.SetNamedValue(clientSecretKey, WDJ::JsonValue::CreateStringValue(windowsTerminalClientSecret));
+        jsonContent.SetNamedValue(refreshTokenKey, WDJ::JsonValue::CreateStringValue(_refreshToken));
+        const auto stringContent = jsonContent.ToString();
+        WWH::HttpStringContent requestContent{
+            stringContent,
+            WSS::UnicodeEncoding::Utf8,
+            applicationJsonString
+        };
+
+        try
+        {
+            const auto jsonResult = _SendRequestReturningJson(accessTokenEndpoint, requestContent, WWH::HttpMethod::Post());
+
+            _authToken = jsonResult.GetNamedString(accessTokenKey);
+            _refreshToken = jsonResult.GetNamedString(refreshTokenKey);
+            _httpClient.DefaultRequestHeaders().Authorization(WWH::Headers::HttpCredentialsHeaderValue{ bearerString, _authToken });
+
+            // raise the new tokens so the app can store them
+            Windows::Foundation::Collections::ValueSet authValues{};
+            authValues.Insert(accessTokenKey, Windows::Foundation::PropertyValue::CreateString(_authToken));
+            authValues.Insert(refreshTokenKey, Windows::Foundation::PropertyValue::CreateString(_refreshToken));
+            _AuthChangedHandlers(*this, winrt::make<GithubCopilotAuthenticationResult>(L"", authValues));
+        }
+        CATCH_LOG();
+    }
+
+    WDJ::JsonObject GithubCopilotLLMProvider::_SendRequestReturningJson(std::wstring_view uri, const WWH::IHttpContent& content, WWH::HttpMethod method)
+    {
+        if (!method)
+        {
+            method = content == nullptr ? WWH::HttpMethod::Get() : WWH::HttpMethod::Post();
+        }
+
+        WWH::HttpRequestMessage request{ method, Uri{ uri } };
+        request.Content(content);
+
+        const auto response{ _httpClient.SendRequestAsync(request).get() };
+        const auto string{ response.Content().ReadAsStringAsync().get() };
```

Comment:
This should ideally use `co_await` so that it doesn't block a worker thread for the entire duration. You can return one of the several task objects, like cppwinrt's `IAsyncOperation<WDJ::JsonObject>` or WIL's `wil::task<WDJ::JsonObject>`. The latter is explained here: https://github.com/microsoft/wil/blob/master/include/wil/coroutine.h
Personally, I think `IAsyncOperation<WDJ::JsonObject>` would be better right now purely because we use it in a bunch of places already.

`void`-functions like `_refreshAuthTokens()` would need to be changed to return `IAsyncAction`.

Reply:
(used `safe_void_coroutine` instead) Fixed!

Reply:
I don't think that's the right type to use, since that coroutine type can't be awaited but it seems like you need to so that the order of operations is correct in the callers, no?

Reply:
Good catch, updated